woocommerce/google-listings-and-ads
# Product Feed UI with API

### Changes proposed in this Pull Request:

Closes #481 .

This PR calls the Product Feed API (developed in PR https://github.com/woocommerce/google-listings-and-ads/pull/513) and displays the product feed in the UI. It supports pagination (10 items per page) and sorting.

#### Technical note: 

- I came up with a new approach in wp-data to support pagination. The paginated list of items is stored under a node called `pages` with page number as key. When users change the sorting, we can easily discard the old `pages` and construct a new one. We don't need to use `JSON.stringify` to create key, or to use `rememo`'s `createSelector` like in [`getMCIssues` selector](https://github.com/woocommerce/google-listings-and-ads/blob/62df13d8d1b421c66a1574ec6fb056f889020a11/js/src/data/selectors.js#L73-L95).

![image](https://user-images.githubusercontent.com/417342/117579190-6d8f9f00-b124-11eb-82d4-2f789b9f16aa.png)

- I also make use of `shouldInvalidate` in resolvers. When users change sorting, we want to invalidate previously resolved data.
- If the above looks okay, I would like to implement the same for `getMCIssues` too later.
- The `ProductFeedTableCard` is similar to the implementation in [`IssuesTableCard`](https://github.com/woocommerce/google-listings-and-ads/blob/62df13d8d1b421c66a1574ec6fb056f889020a11/js/src/product-feed/issues-table-card/index.js#L98-L175). I am thinking to refactor them into [`external-components`](https://github.com/woocommerce/google-listings-and-ads/blob/trunk/js/src/external-components/README.md), in a separate PR later.

### Screenshots:

Demo video with my voice (https://d.pr/v/qeR8JZ):

https://user-images.githubusercontent.com/417342/117579032-b6932380-b123-11eb-9532-8674cca147ae.mov

### Detailed test instructions:

1. Open https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fproduct-feed
2. Scroll down to the Product Feed section.
3. Test the pagination and sorting feature. Make sure the products are listed as per pagination and sort expectation.

### Changelog Note:

Display product feed in the UI.
